src/Mailgun/Mailgun.php

```diff
@@ -47,6 +60,51 @@ public function __construct(
     ) {
         $this->apiKey = $apiKey;
         $this->restClient = new RestClient($apiKey, $apiEndpoint, $httpClient);
+
+        $this->initializeRegionClients($httpClient);
+    }
+
+    /**
+     * This function creates clients for each individual region. Messages are
+     * only accessible from the region that received them, so having a mapping
+     * of region -> client seems reasonable.
+     *
+     * @param HttpClient|null $httpClient
+     *
+     * @return bool
+     */
+    private function initializeRegionClients(HttpClient $httpClient = null)
+    {
+        foreach ($this->regions as $region => $apiUrl) {
+            $this->regionClients[$region] = new RestClient(
+                $this->apiKey,
+                $apiUrl,
+                $httpClient
+            );
+        }
+
+        return true;
+    }
+
+    /**
+     * Tries to select the correct API region for a URL through simple
+     * comparisons with the known region URLs. If none are matched, null is
+     * returned.
+     *
+     * @param string $url
+     *
+     * @return \string
+     */
+    private function regionForUrl($url)
+    {
+        foreach ($this->regions as $region => $apiUrl) {
+            if (strpos($url, 'https://' . $apiUrl) === 0) {
+                return $region;
+            }
+        }
+
+        // Region in URL does *not* exist in the regions mapping.
+        return null;
```

Comment:
As discussed on my previous review, `return;` is the PSR standard for `return null;`

Also, I can see the `return @\string` annotation, but if I am not wrong `\string` is part of PHP7 annotations, isn't it? So I think a better approach here is `@return string` to be consistent with PHP5 code. In addition, given this method can return `null` too, it should be added to the annotation.